### PR TITLE
fix wide variance in success rate

### PR DESCRIPTION
**Pull request by dyth**
The success rate computed by `scripts/evaluate.py` has a wide variance with respect to the batch size (`num_envs` evaluated per time).

This is caused by variable-length instruction padding. The variable length changes the output from a unidirectional GRU. Padding should not change the embedding value.

This was fixed by taking the output of the GRU computed before the first zero in the input sequence appears (pad value).

### CODE REVIEW

**Comment by dyth** [comment]
*TL;DR there is still some variance due to randomness in the GPU, but is statistically insignificant*

null hypothesis: distribution of returns with batch size 32 and 1 are identical
alternate: distributions are non-identical

thus use Two-Tailed Independent Two-Sample T-Test with equal sample size (1000)
assume same underlying population variance

when batch size = 32; x, sigma : 0.483 0.416
when batch size = 1; x, sigma: 0.470 0.422

t = (0.483 - 0.470) / (sqrt((0.416 ** 2 + 0.422 ** 2) / 2) * sqrt(2 / 1000)) = 0.689
Two-tailed test, thus t is 0.344

dof = 2 * 1000 - 2 = 1998
significance value is 0.49 %

0.49 > 0.344, thus the result is not significant at p < 0.1
.: no reason not to accept null hypothesis

This randomness is caused by tiny variations in the initialisation of `agent.memory` in https://github.com/mila-iqia/babyai/blob/756df006e210295560dab0f2f166a06ae46ccafa/babyai/utils/agent.py#L53

File: babyai/model.py [hunk 1]
```diff
@@ -255,8 +255,10 @@ def forward(self, obs, memory, instr_embedding=None):
 
     def _get_instr_embedding(self, instr):
         if self.lang_model == 'gru':
-            _, hidden = self.instr_rnn(self.word_embedding(instr))
-            return hidden[-1]
+            out, _ = self.instr_rnn(self.word_embedding(instr))
+            index = -1 - (instr <= 0).sum(dim=1)
+            hidden = out[range(len(index)), index, :]
```
Comment (rizar) [inline babyai/model.py]: if you could please use the same code in this `if` branch and in the next one for computing `lengths`, or `index`, that would be great. The same job is done in both cases, it would be great to avoid code duplication.
